pkg/intel/me/me.go

```diff
@@ -0,0 +1,280 @@
+	fmt.Fprintf(&b, " TicksToAdd         : 0x%x\n", h.TicksToAdd)
+	fmt.Fprintf(&b, " TokensToAdd        : 0x%x\n", h.TokensToAdd)
+	fmt.Fprintf(&b, " UMASizeOrReserved  : 0x%x\n", h.UMASizeOrReserved)
+	fmt.Fprintf(&b, " FlashLayoutOrFlags : 0x%x\n", h.FlashLayoutOrFlags)
+	fmt.Fprintf(&b, " Fitc Version       : %d.%d.%d.%d\n", h.FitcMajor, h.FitcMinor, h.FitcHotfix, h.FitcBuild)
+
+	return b.String()
+}
+
+// FlashPartitionTableEntry describes information of a flash partition table entry.
+type FlashPartitionTableEntry struct {
+	Name           [4]uint8
+	Owner          [4]uint8
+	Offset         uint32
+	Length         uint32
+	StartTokens    uint32
+	MaxTokens      uint32
+	ScratchSectors uint32
+	Flags          uint32
+}
+
+func (e FlashPartitionTableEntry) String() string {
+
+	var b strings.Builder
+
+	b.WriteString("Flash partition entry:\n")
+	fmt.Fprintf(&b, " Name          : %s\n", []byte{e.Name[0], e.Name[1], e.Name[2], e.Name[3]})
+	fmt.Fprintf(&b, " Owner         : %s\n", []byte{e.Owner[0], e.Owner[1], e.Owner[2], e.Owner[3]})
+	fmt.Fprintf(&b, " Offset        : 0x%x\n", e.Offset)
+	fmt.Fprintf(&b, " Length        : 0x%x\n", e.Length)
+	fmt.Fprintf(&b, " StartTokens   : 0x%x\n", e.StartTokens)
+	fmt.Fprintf(&b, " MaxTokens     : 0x%x\n", e.MaxTokens)
+	fmt.Fprintf(&b, " ScratchSectors: 0x%x\n", e.ScratchSectors)
+	fmt.Fprintf(&b, " Flags         : 0x%x\n", e.Flags)
+
+	if e.Flags>>24 == 0xff {
+		b.WriteString(" Valid         : No\n")
+	} else {
+		b.WriteString(" Valid         : yes\n")
+	}
+	if e.Flags&1 > 0 {
+		b.WriteString(" Partition     : Data\n")
+	} else {
+		b.WriteString(" Partition     : Code\n")
+	}
+
+	return b.String()
+}
+
+// IntelME abstracts the ME/CSME/SPS firmware found on intel platforms
+type IntelME struct {
+	hdr        FlashPartitionTableHeader
+	legacyhdr  LegacyFlashPartitionTableHeader
+	legacy     bool
+	partitions []FlashPartitionTableEntry
+	image      []byte
+	// Offset in image to $FPT
+	fptoffset uint32
+}
+
+// ParseIntelFirmware parses the Intel firmware image by uefi.Firmware interface`
+func ParseIntelFirmware(firmware uefi.Firmware) (*IntelME, error) {
+	uefi, err := ParseIntelFirmwareBytes(firmware.Buf())
+	if err != nil {
+		return nil, fmt.Errorf("unable to get the content of file: %v", err)
+	}
+
+	return uefi, nil
+}
+
+// ParseIntelFirmwareBytes parses the Intel firmware image from bytes
+func ParseIntelFirmwareBytes(imageBytes []byte) (me *IntelME, err error) {
+
+	legacy := false
+	fptoffset := -1
+	// Search for the Flash partition table
+	for i := 0; i < len(imageBytes); i += 0x1000 {
+
+		// New Header
+		if bytes.HasPrefix(imageBytes[i:], []byte(`$FPT`)) {
+			fptoffset = i
+			break
+		}
+		// Legacy Header
+		if bytes.HasPrefix(imageBytes[i:], append(make([]byte, 16), []byte(`$FPT`)...)) {
+			legacy = true
+			fptoffset = i
+			break
+		}
+	}
+	if fptoffset == -1 {
+		err = fmt.Errorf("no FlashPartitionTable found")
+		return
```

Comment:
Just do a `return nil, fmt.Errorf("no FlashPartitionTable found")` here and elsewhere. See https://tour.golang.org/basics/7 "Naked return statements should be used only in short functions, as with the example shown here. They can harm readability in longer functions."